mittab/apps/tab/tab_card.py

```diff
@@ -0,0 +1,284 @@
+        "wings": [judge.name for judge in round_obj.judges.exclude(pk=round_obj.chair.pk)]
+    }
+    
+    opponent = round_obj.opp_team if round_obj.gov_team == team else round_obj.gov_team
+    if opponent:
+        opponent_debaters = list(opponent.debaters.all())
+        jsonRound["opponent"] = {
+            "name": opponent.get_team_code(),
+            "school": opponent.school.name,
+            "debater1": opponent_debaters[0].name if opponent_debaters else None,
+            "debater2": opponent_debaters[1].name if len(opponent_debaters) > 1 else None,
+        }
+
+    jsonRound["debater1"] = list(
+        (stat.speaks, stat.ranks) 
+        for stat in RoundStats.objects.filter(debater=deb1, round=round_obj)
+    )
+    jsonRound["debater2"] = list(
+        (stat.speaks, stat.ranks) 
+        for stat in RoundStats.objects.filter(debater=deb2, round=round_obj)
+    ) if deb2 else []
+
+    try:
+        bye_round = Bye.objects.get(bye_team=team).round_number
+        jsonRound[bye_round - 1][bye_round]
+    except Bye.DoesNotExist:
+        pass
+
+    return jsonRound
+
+class JSONDecimalEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, Decimal):
+            return float(obj)
+        return json.JSONEncoder.default(self, obj)
+    
+def get_all_json_data():
+    all_tab_cards_data = {}
+    
+    # Prefetch related data to reduce database queries
+    teams = Team.objects.prefetch_related(
+        'school',
+        Prefetch('debaters', queryset=Debater.objects.all()),
+        Prefetch('gov_team', queryset=Round.objects.all()),
+        Prefetch('opp_team', queryset=Round.objects.all())
+    )
+
+    total_rounds = TabSettings.objects.get(key="tot_rounds").value
+
+    for team in teams:
+        tab_card_data = {
+            "team_name": team.get_team_code(),
+            "team_school": team.school.name,
+            "rounds": [{}] * total_rounds
+        }
+
+        debaters = list(team.debaters.all())
+        deb1 = debaters[0]
+        tab_card_data["debater_1"] = deb1.name
+        tab_card_data["debater_1_status"] = Debater.NOVICE_CHOICES[deb1.novice_status][1]
+        
+        if len(debaters) > 1:
+            deb2 = debaters[1]
+            tab_card_data["debater_2"] = deb2.name
+            tab_card_data["debater_2_status"] = Debater.NOVICE_CHOICES[deb2.novice_status][1]
+        else:
+            deb2 = None
+
+        rounds = list(team.gov_team.all()) + list(team.opp_team.all())
+
+        for round_obj in rounds:
+            tab_card_data["rounds"][round_obj.round_number - 1] = json_get_round(round_obj, team, deb1, deb2)
+        
+        all_tab_cards_data[team.get_team_code()] = tab_card_data
+    return all_tab_cards_data
+
+def get_tab_card_data(request, team_id):
+    try:
+        team_id = int(team_id)
+    except ValueError:
+        return redirect_and_flash_error(request, "Invalid team id")
+    
+    team = Team.objects.get(pk=team_id)
```

Comment:
`prefetch_related` let's us avoid the N+1 problem: https://stackoverflow.com/questions/97197/what-is-the-n1-selects-problem-in-orm-object-relational-mapping

Essentially: queries to SQL are (relatively) computationally expensive, they require network traffic, can hit SQL locks, etc. So unless there's a particularly good reason not to, we should take advantage of some Django APIs that let us get all the data we want in 1 query instead of accidentally creating more

```
# opp_team fields = rounds as opp_team, gov_team field = rounds as gov_team
team = Team.objects.get(pk=team_id).prefetch_related("gov_team", "opp_team", "debaters")
rounds = list(team.gov_team) + list(team.opp_team)
debaters = list(team.debaters)
```